DEVELOPERS.md

```diff
@@ -0,0 +1,165 @@
+
+* [Development Setup](#setup)
+* [Coding Rules](#rules)
+* [Commit Message Guidelines](#commits)
+* [Troubleshooting](#troubleshooting)
+
+## <a name="setup"></a> Development Setup
+
+This document describes how to set up your development environment to build and test Iroh, and
+explains the basic mechanics of using `cargo run` and `cargo test`.
+
+### Installing Dependencies
+
+Before you can build Iroh, you must install and configure the following dependencies on your
+machine:
+
+* [Git](http://git-scm.com/): The [Github Guide to
+  Installing Git][git-setup] is a good source of information.
+
+* [The Rust Programming Language](https://www.rust-lang.org/): see https://www.rust-lang.org/learn/get-started to get started
+
+
+### Forking Iroh on Github
+
+To contribute code to Iroh, you must have a GitHub account so you can push code to your own
+fork of Iroh and open Pull Requests in the [GitHub Repository][github].
+
+
+### Building Iroh from source
+
+Use `cargo` commands to build and run the various Iroh binaries.
+For example:
+```shell
+# run each command in a different terminal to simulate running iroh as
+# microservices on different boxes:
+$ cargo run -p iroh-p2p
+$ cargo run -p iroh-gateway
+$ cargo run -p iroh-store
+$ cargo run -p iroh-ctl -- status --watch
+```
+
+You also need a recent version of the [Protobuf compiler](https://github.com/protocolbuffers/protobuf#protocol-compiler-installation).
+
+Download it from the [Protobuf Releases page](https://github.com/protocolbuffers/protobuf/releases); you need to get the `protoc-` release for your platform. To install, make sure the `protoc` compiler is on your path. If you get errors during build about `experimental_allow_proto3_optional` or inability to import `/google/protobuf/empty.proto` you're likely using a version of the compiler that's too old.
+
+## <a name="rules"></a> Coding Rules
+
+When you push your branch to github and open up a pull request, it will automatically trigger  [CircleCI](https://circleci.com/about/) to lint and test your code.
+
+In order to catch linting and testing errors before pushing the code to github, be sure to run:
+
+```shell
+$ cargo clippy --workspace --examples --benches --tests
+$ cargo test --workspace --examples --benches
+```
+
+Setting up a [git hook][git-hook] to run these commands can save you many headaches:
+```shell
+#!/bin/sh
+cargo clippy --workspace --examples --tests --benches && cargo test --workspace --examples
+```
+
+## <a name="Pull Requests"></a> Pull Request Guidelines
+The tests must pass and you must get an approval from someone on the Iroh team before you can merge your PR.
+
+Depending on your permissions in the `iroh` repo, you may not have the the ability to "request a review". Instead, please tag your selected reviewers in the PR itself (using the `@`) and specify that you would like them to review. If you are apart of our discord community, you can and should ping your reviewer(s) there as well.
```

Comment:
It would be helpful here perhaps to explain how to even find potential reviewers. Depending on the involvement of the contributor they may not know their github name.

Reply:
great point